OpenRA.Mods.Cnc/ImportTiberianDawnLegacyMapCommand.cs

```diff
@@ -0,0 +1,161 @@
+
+		[Desc("FILENAME", "Convert a legacy Tiberian Dawn INI/MPR map to the OpenRA format.")]
+		public override void Run(ModData modData, string[] args)
+		{
+			base.Run(modData, args);
+		}
+
+		public override int MapSize
+		{
+			// NOTE: original engine limitation
+			get { return 64; }
+		}
+
+		public override void ValidateMapFormat(int format)
+		{
+			if (format > 1)
+			{
+				Console.WriteLine("ERROR: Detected NewINIFormat {0}. Are you trying to import a Red Alert map?".F(format));
+				return;
+			}
+		}
+
+		static Dictionary<string, Pair<byte, byte>> overlayResourceMapping = new Dictionary<string, Pair<byte, byte>>()
+		{
+			// Tiberium
+			{ "ti1", new Pair<byte, byte>(1, 0) },
+			{ "ti2", new Pair<byte, byte>(1, 1) },
+			{ "ti3", new Pair<byte, byte>(1, 2) },
+			{ "ti4", new Pair<byte, byte>(1, 3) },
+			{ "ti5", new Pair<byte, byte>(1, 4) },
+			{ "ti6", new Pair<byte, byte>(1, 5) },
+			{ "ti7", new Pair<byte, byte>(1, 6) },
+			{ "ti8", new Pair<byte, byte>(1, 7) },
+			{ "ti9", new Pair<byte, byte>(1, 8) },
+			{ "ti10", new Pair<byte, byte>(1, 9) },
+			{ "ti11", new Pair<byte, byte>(1, 10) },
+			{ "ti12", new Pair<byte, byte>(1, 11) },
+		};
+
+		void UnpackTileData(Stream ms)
+		{
+			for (var j = 0; j < MapSize; j++)
+			{
+				for (var i = 0; i < MapSize; i++)
+				{
+					var type = ms.ReadUInt8();
+					var index = ms.ReadUInt8();
+					Map.MapTiles.Value[new CPos(i, j)] = new TerrainTile(type, index);
+				}
+			}
+		}
+
+		void ReadOverlay(IniFile file)
+		{
+			var overlay = file.GetSection("OVERLAY", true);
+			if (overlay == null)
+				return;
+
+			foreach (var kv in overlay)
+			{
+				var loc = Exts.ParseIntegerInvariant(kv.Key);
+				var cell = new CPos(loc % MapSize, loc / MapSize);
+
+				var res = Pair.New((byte)0, (byte)0);
+				var type = kv.Value.ToLowerInvariant();
+				if (overlayResourceMapping.ContainsKey(type))
+					res = overlayResourceMapping[type];
+
+				Map.MapResources.Value[cell] = new ResourceTile(res.First, res.Second);
+				if (OverlayActors.Contains(type))
+				{
+					var ar = new ActorReference(type)
+					{
+						new LocationInit(cell),
+						new OwnerInit("Neutral")
+					};
+
+					var actorCount = Map.ActorDefinitions.Count;
+					Map.ActorDefinitions.Add(new MiniYamlNode("Actor" + actorCount++, ar.Save()));
+				}
+			}
+		}
+
+		void ReadTrees(IniFile file)
+		{
+			var terrain = file.GetSection("TERRAIN", true);
+			if (terrain == null)
+				return;
+
+			foreach (var kv in terrain)
+			{
+				var loc = Exts.ParseIntegerInvariant(kv.Key);
+				var ar = new ActorReference(kv.Value.Split(',')[0].ToLowerInvariant())
+				{
+					new LocationInit(new CPos(loc % MapSize, loc / MapSize)),
+					new OwnerInit("Neutral")
+				};
+
+				var actorCount = Map.ActorDefinitions.Count;
+				Map.ActorDefinitions.Add(new MiniYamlNode("Actor" + actorCount++, ar.Save()));
+			}
+		}
+
+		public override void LoadPlayer(IniFile file, string section)
```

Comment:
You could drop duplication here by defining mapping dictionaries that are referred to from the base class.